zebROS_ws/src/tf_object_detection/src/object_detection.cpp

```diff
@@ -26,30 +29,288 @@ void camera_info_callback(const sensor_msgs::CameraInfoConstPtr &info)
 {
-	double sum = 0.0;
-	unsigned count = 0;
-	for (int j = bound_rect.tl().y+1; j < bound_rect.br().y; j++) //for each row
-	{
-		const float *ptr_depth = depth.ptr<float>(j);
+	// setup randomizing (for initialization of k-means)
+	std::random_device seeder;
+	std::mt19937 engine(seeder());
+
+	// Start of checking if k is too high
 
-		for (int i = bound_rect.tl().x+1; i < bound_rect.br().x; i++) //for each pixel in row
+	// Convert cropped depth matrix to vector
+	std::vector<float> vec;
+	cv::Mat mat = depth(bound_rect);
+	for (int i = 0; i < mat.rows; i++) {
+    vec.insert(vec.end(), mat.ptr<float>(i), mat.ptr<float>(i)+mat.cols*mat.channels());
+  }
+
+	// Find unique (defined as >1cm away) values in the vector
+	std::sort(vec.begin(), vec.end());
+	auto last = std::unique(vec.begin(), vec.end(), [](float first, float second){ return fabs(second - first)<0.01; });
+	vec.erase(last, vec.end());
+
+	// Set k to maximumK if there are more than maximumK unique values, otherwise set it to the number of unique values
+	int k = std::min((int)(vec.size()), maximumK);
+
+	// End of checking if k is too high
+
+	// initialize arrays (and a vector) for k-means
+	double centroids[k];
+	double prevCentroids[k];
+	std::vector<float> clusters[k];
+
+	// Pick centroids
+	std::sample(vec.begin(), vec.end(), centroids, k, engine);
+
+	// Print centroids
+	if (debug) {
+		ROS_INFO_STREAM("k = " << k);
+		ROS_INFO_STREAM("Centroids: ");
+		for (int i = 0; i < k; i++) {
+			ROS_INFO_STREAM(centroids[i]);
+		}
+	}
+
+	while (true) { // once the algorithm converges this returns
+		for (int j = bound_rect.tl().y+1; j < bound_rect.br().y; j++) // for each row
 		{
-			if (!(isnan(ptr_depth[i]) || isinf(ptr_depth[i]) || (ptr_depth[i] <= 0)))
+			const float *ptr_depth = depth.ptr<float>(j);
+
+			for (int i = bound_rect.tl().x+1; i < bound_rect.br().x; i++) // for each pixel in row
 			{
-				sum += ptr_depth[i];
-				count += 1;
+				if (!(isnan(ptr_depth[i]) || isinf(ptr_depth[i]) || (ptr_depth[i] <= 0)))
+				{
+					// Calculate which centroid/mean the current pixel is closest to
+					float diffs[k];
+					for (int c = 0; c < k; c++) {
+						diffs[c] = abs(centroids[c] - ptr_depth[i]);
+					}
+					int closestCentroid = std::distance(diffs, std::min_element(diffs, diffs+k));
+					// Append the pixel's value to the cluster corresponding to that centroid
+					clusters[closestCentroid].push_back(ptr_depth[i]);
+				}
+			}
+		}
+
+		// Recalculate centroids using the average of the cluster closest to each centroid
+		// (or set the centroid to 999 if there are no values in the cluster)
+		for (int i = 0; i < k; i++) {
+			double sum = 0;
+			for (float f : clusters[i]) {
+				sum += f;
 			}
+			if (clusters[i].size() != 0) {
+				centroids[i] = sum / (double)clusters[i].size();
+			} else { // If the centroid's cluster has no values, set it to 999 (basically remove it)
+				centroids[i] = 999;
```

Comment:
Could use numeric_limits<>::max() here?

Reply:
Good idea -- I changed that (and other instances of `999`, I used it in the `contoursDepthMat` function too) to `std::numeric_limits<float>::max()`